server.js

```diff
@@ -1,20 +1,119 @@
 const express = require("express");
 const cors = require("cors");
-
+const moment = require("moment"); // require
+moment().format("YYYY-MM-DD");
 const app = express();
 
 app.use(express.json());
 app.use(cors());
 
 //Use this array as your (in-memory) data store.
 const bookings = require("./bookings.json");
+const { body, validationResult } = require("express-validator");
 
 app.get("/", function (request, response) {
   response.send("Hotel booking server.  Ask for /bookings, etc.");
 });
 
 // TODO add your routes and helper functions here
+app.post(
+  "/bookings",
+  [body("email").isEmail().withMessage("Please input a valid email")],
+  function (request, response) {
+    const errors = validationResult(request);
+    if (!errors.isEmpty()) {
+      return response.status(422).json({ errors: errors.array() });
+    }
+    const expectedKeys = [
+      "title",
+      "firstName",
+      "surname",
+      "email",
+      "roomId",
+      "checkInDate",
+      "checkOutDate",
+    ];
+    for (let item in request.body) {
+      if (!request.body[item] || !expectedKeys.includes(item)) {
+        response
+          .status(400)
+          .json("Please ensure your are passing the correct fileds");
+        return;
+      }
+    }
+    if (!moment(request.body.checkOutDate).isAfter(request.body.checkInDate)) {
+      response.status(422).json("Checkout date must be after checkin date");
+      return;
+    }
+    let newBooking = {
+      id: bookings[bookings.length - 1].id + 1,
+      title: request.body.title,
+      firstName: request.body.firstName,
+      surname: request.body.surname,
+      email: request.body.email,
+      roomId: request.body.roomId,
+      checkInDate: request.body.checkInDate,
+      checkOutDate: request.body.checkOutDate,
+    };
+    bookings.push(newBooking);
+    response.json(bookings);
+  }
+);
+
+app.get("/bookings", function (request, response) {
+  response.send(bookings);
+});
+
+app.get("/bookings/search", function (request, response) {
+  let searchTerm = request.query.term;
+  let date = request.query.date;
+  if (moment(date, "YYYY-MM-DD", true).isValid()) {
+    let results = bookings.filter(
+      (booking) =>
+        date === booking.checkInDate ||
+        moment(date).isBetween(booking.checkInDate, booking.checkOutDate)
+    );
+    results.length > 0
+      ? response.json(results)
+      : response.status(404).json("nothing was found on this date");
+  } else if (typeof searchTerm == "string" && searchTerm.trim().length > 0) {
+    const result = bookings.filter(
+      (booking) =>
+        booking.email.toLowerCase().includes(searchTerm.toLowerCase()) ||
+        booking.firstName.toLowerCase().includes(searchTerm.toLowerCase()) ||
+        booking.surname.toLowerCase().includes(searchTerm.toLowerCase())
+    );
```

Comment:
It would be more efficient to assign searchTerm.toLowerCase() to a variable up front, and use this in the filter

Reply:
I tried that actually, but I was giving me an error saying that it can't use this function on an undefined. specially if i'm querying dy date and there is no term passed in the request.

Reply:
I can't work out how to recreate this error. But if it is happening because the searchTerm is undefined, you could add a check for that.